auth/common.cc

```diff
@@ -113,4 +118,118 @@ ::service::query_state& internal_distributed_query_state() noexcept {
     return qs;
 }
 
+future<::service::group0_guard> start_group0_operation(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        seastar::abort_source* as) {
+    co_return co_await qp.container().invoke_on(0,
+        [&group0_client, as](cql3::query_processor& qp) -> future<::service::group0_guard> {
+            co_return co_await group0_client.start_operation(as);
+    });
+}
+
+static future<> announce_mutations_with_guard(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    co_await qp.container().invoke_on(0,
+        [&group0_client, muts = std::move(muts), group0_guard = std::move(group0_guard), as]
+        (cql3::query_processor&) mutable -> future<> {
+            auto group0_cmd = group0_client.prepare_command(
+                ::service::write_mutations{
+                    .mutations{std::move(muts)},
+                },
+                group0_guard,
+                "auth: modify internal data"
+            );
+            co_await group0_client.add_entry(std::move(group0_cmd), std::move(group0_guard), as);
+    });
+}
+
+static future<> announce_mutations_maybe_split(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    auto begin = muts.begin();
+    auto end = muts.end();
+    size_t batch_size  = muts.size();
+    auto remaining = batch_size;
+    while (remaining) {
+        try {
+            co_await announce_mutations_with_guard(qp, group0_client, {begin, end}, std::move(group0_guard), as);
+            remaining -= (end - begin);
+            begin = end;
+        } catch (raft::command_is_too_big_error&) {
+            batch_size /= 2;
+            if (batch_size == 0) {
+                std::throw_with_nested(std::runtime_error("raft add entry failed with batch size 1"));
+            }
+            // when retrying we don't change begin and end gets truncated below
+        }
```

Comment:
I don't know exactly command overhead and in general estimating here,  because raft asserts size after ser::serialize and that's potentially not the same as sum of mut.representation().size();

Reply:
In case of cdc over raft, we just assume that having `max_command_size / 2` as the limit will account for the overhead (which should be bounded by some constant):

```c++
const size_t max_command_size = _raft.max_command_size();
const size_t mutation_size_threshold = max_command_size / 2;
auto gen_mutations = co_await cdc::get_cdc_generation_mutations_v3(
    gen_table_schema, gen_uuid, gen_desc, mutation_size_threshold, guard.write_timestamp());
```

I think it would be OK to assume the same here as well as it will lead to simpler code.

Reply:
Potentially yes, but such assumption lead to more fragile code because it introduces hidden invariant and when it won't hold it will break the code.

As a consequence you need to be more conservative then (e.g. using only half of max_command_size) and then it would need to potentially do more unnecessary splitting.

Reply:
OK. I wish there were a way to calculate the serialized size so that we don't have to take care of the exception, but implementing something like this is out of scope so I suppose the current code is fine as it is right now.

Reply:
`max_command_size` is part of raft server's interface.
So raft server should guarantee that if you provide a command of size below this threshold, it will pass.
If it doesn't -- that's a bug in raft server implementation, and we should catch such bugs as early as possible, not swipe them under the rug with retry loops such as this.

In other words I think we should get rid of the retry loop.

Whether you multiply by 1/2 or by 3/4 most likely doesn't matter.

> As a consequence you need to be more conservative then (e.g. using only half of max_command_size) and then it would need to potentially do more unnecessary splitting.

Meh, it's still large enough to handle deleting thousands of roles. Probably tens or hundreds of thousands. Have you attempted a back-of-the-envelope estimation?

Reply:
I wrote about this already, it's not about back-of-the-envelope estimation but it's about silently coding dependencies - here dependency between mutation(s) size and max_command_size. Althoguh I don't insist it's always easier to remove the code than add it, so I can remove catch (raft::command_is_too_big_error&)  handling